File: android/app/src/main/java/com/claudelists/app/api/AuthManager.kt
package com.claudelists.app.api

import android.content.Context
import android.content.Intent
import android.util.Log
import com.claudelists.app.BuildConfig
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.tasks.await
import kotlinx.serialization.json.*
import java.security.MessageDigest

private const val TAG = "AuthManager"

/**
 * Convert Google ID to a deterministic UUID (matches backend auth.ts logic).
 * Uses SHA-256 hash formatted as UUID v4 format.
 */
private fun googleIdToUuid(googleId: String): String {
    val data = "google:$googleId".toByteArray(Charsets.UTF_8)
    val digest = MessageDigest.getInstance("SHA-256")
    val hashBytes = digest.digest(data)

    // Take first 16 bytes and format as UUID
    val hex = hashBytes.take(16).joinToString("") { "%02x".format(it) }

    // Format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
    // Set version (4) and variant bits
    val byte6 = (hashBytes[6].toInt() and 0x0f) or 0x40 // version 4
    val byte8 = (hashBytes[8].toInt() and 0x3f) or 0x80 // variant

    return "${hex.substring(0, 8)}-${hex.substring(8, 12)}-4${hex.substring(13, 16)}-${"%02x".format(byte8)}${hex.substring(18, 20)}-${hex.substring(20, 32)}"
}

/**
 * Manages authentication using native Google Sign-In.
 *
 * Uses Google ID token for Edge Function API authentication.
 * Exchanges Google token with Supabase for realtime subscriptions.
 */
class AuthManager(
    private val context: Context,
    private val httpClient: HttpClient
) {

    private val googleWebClientId = BuildConfig.GOOGLE_WEB_CLIENT_ID
    private val apiBaseUrl = BuildConfig.API_BASE_URL
    private val apiAnonKey = BuildConfig.API_ANON_KEY

    private val googleSignInClient: GoogleSignInClient by lazy {
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(googleWebClientId)
            .requestEmail()
            .build()
        GoogleSignIn.getClient(context, gso)
    }

    // Auth state
    private val _authState = MutableStateFlow(loadAuthState())
    val authState: StateFlow<AuthState> = _authState

    // Loading state for UI
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    // Error state for UI
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    // Supabase access token for realtime
    private var supabaseAccessToken: String? = null

    data class AuthState(
        val isAuthenticated: Boolean = false,
        val userId: String? = null,
        val userEmail: String? = null,
        val userName: String? = null
    )

    private fun loadAuthState(): AuthState {
        val account = GoogleSignIn.getLastSignedInAccount(context)
        return if (account != null && account.idToken != null && account.id != null) {
            AuthState(
                isAuthenticated = true,
                userId = googleIdToUuid(account.id!!),
                userEmail = account.email,
                userName = account.displayName
            )
        } else {
            AuthState()
        }
    }

    /**
     * Get the Google ID token for API authentication.
     * Refreshes silently if needed.
     */
    suspend fun getGoogleIdToken(): String? {
        return try {
            // Try silent sign-in to refresh the token
            val account = googleSignInClient.silentSignIn().await()
            account.idToken?.also { Log.d(TAG, "DEBUG_TOKEN: $it") }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get ID token", e)
            // Fall back to last signed in account
            GoogleSignIn.getLastSignedInAccount(context)?.idToken?.also { Log.d(TAG, "DEBUG_TOKEN: $it") }
        }
    }

    /**
     * Get the Supabase access token for realtime subscriptions.
     * Exchanges Google token if needed.
     */
    suspend fun getValidAccessToken(): String? {
        // Return cached token if available
        supabaseAccessToken?.let { return it }

        // Exchange Google token for Supabase session
        val googleToken = getGoogleIdToken() ?: return null
        return exchangeForSupabaseSession(googleToken)
    }

    /**
     * Exchange Google ID token for Supabase session.
     */
    private suspend fun exchangeForSupabaseSession(googleIdToken: String): String? {
        return try {
            val response = httpClient.post("$apiBaseUrl/auth/v1/token") {
                parameter("grant_type", "id_token")
                contentType(ContentType.Application.Json)
                header("apikey", apiAnonKey)
                setBody(buildJsonObject {
                    put("provider", "google")
                    put("id_token", googleIdToken)
                }.toString())
            }

            if (response.status.isSuccess()) {
                val json = Json.parseToJsonElement(response.bodyAsText()).jsonObject
                val accessToken = json["access_token"]?.jsonPrimitive?.content
                if (accessToken != null) {
                    supabaseAccessToken = accessToken
                    Log.i(TAG, "Supabase session obtained")
                }
                accessToken
            } else {
                Log.e(TAG, "Supabase token exchange failed: ${response.status}")
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Supabase token exchange error", e)
            null
        }
    }

    fun getUserId(): String? = _authState.value.userId

    /**
     * Get the sign-in intent to launch.
     * Call this and launch the intent with an ActivityResultLauncher.
     */
    fun getSignInIntent(): Intent {
        _isLoading.value = true
        _error.value = null
        return googleSignInClient.signInIntent
    }

    /**
     * Handle the result from the sign-in intent.
     * Call this from the ActivityResultCallback.
     */
    fun handleSignInResult(data: Intent?): Boolean {
        return try {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            val account = task.getResult(ApiException::class.java)

            if (account.idToken != null && account.id != null) {
                _authState.value = AuthState(
                    isAuthenticated = true,
                    userId = googleIdToUuid(account.id!!),
                    userEmail = account.email,
                    userName = account.displayName
                )
                Log.i(TAG, "Signed in: ${account.email}")
                _isLoading.value = false
                true
            } else {
                Log.e(TAG, "No ID token received")
                _error.value = "No ID token received"
                _isLoading.value = false
                false
            }
        } catch (e: ApiException) {
            Log.e(TAG, "Google Sign-In failed: ${e.statusCode}", e)
            _error.value = "Sign-in failed: ${e.statusCode}"
            _isLoading.value = false
            false
        }
    }

    /**
     * Sign out and clear state.
     */
    fun signOut() {
        googleSignInClient.signOut()
        supabaseAccessToken = null
        _authState.value = AuthState()
    }

    fun clearError() {
        _error.value = null
    }
}

File: ios/CourtLists/CourtLists/Services/AuthManager.swift
import Foundation
import GoogleSignIn
import CryptoKit

@MainActor
class AuthManager: ObservableObject {
    @Published var currentUser: User?
    @Published var isLoading = true
    @Published var error: String?

    /// The Google ID token used for API authentication
    private(set) var idToken: String?

    /// Supabase access token for realtime subscriptions
    private(set) var supabaseAccessToken: String?

    /// UUID derived from Google ID (matches backend format)
    private(set) var userUuid: String?

    init() {
        restoreSession()
    }

    private func restoreSession() {
        GIDSignIn.sharedInstance.restorePreviousSignIn { [weak self] user, error in
            Task { @MainActor in
                if let user = user {
                    await self?.setUser(from: user)
                }
                self?.isLoading = false
            }
        }
    }

    func signIn(presenting: UIViewController) {
        let config = GIDConfiguration(clientID: Configuration.googleClientId)
        GIDSignIn.sharedInstance.configuration = config

        GIDSignIn.sharedInstance.signIn(withPresenting: presenting) { [weak self] result, error in
            Task { @MainActor in
                if let error = error {
                    print("Sign-in error: \(error.localizedDescription)")
                    self?.error = error.localizedDescription
                    return
                }

                guard let user = result?.user else {
                    self?.error = "Failed to get user"
                    return
                }

                await self?.setUser(from: user)
            }
        }
    }

    private func setUser(from googleUser: GIDGoogleUser) async {
        // Store the ID token for API calls
        idToken = googleUser.idToken?.tokenString

        // Compute UUID from Google ID (matches backend format)
        let googleId = googleUser.userID ?? ""
        userUuid = googleIdToUuid(googleId)

        // Create user from Google profile
        currentUser = User(
            id: googleId,
            email: googleUser.profile?.email,
            displayName: googleUser.profile?.name ?? "User",
            photoUrl: googleUser.profile?.imageURL(withDimension: 100)
        )

        print("User signed in: \(currentUser?.email ?? "unknown"), uuid: \(userUuid ?? "none")")

        // Exchange Google token for Supabase session (for realtime)
        if let googleIdToken = idToken {
            await exchangeForSupabaseSession(googleIdToken: googleIdToken)
        }
    }

    /// Refresh the ID token if needed (called before API requests)
    func refreshTokenIfNeeded() async {
        guard let user = GIDSignIn.sharedInstance.currentUser else { return }

        do {
            try await user.refreshTokensIfNeeded()
            await MainActor.run {
                self.idToken = user.idToken?.tokenString
            }
        } catch {
            print("Token refresh failed: \(error)")
        }
    }

    /// Exchange Google ID token for Supabase session (for realtime)
    private func exchangeForSupabaseSession(googleIdToken: String) async {
        let url = URL(string: "\(Configuration.apiBaseUrl)/auth/v1/token?grant_type=id_token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(Configuration.apiAnonKey, forHTTPHeaderField: "apikey")

        let body: [String: Any] = [
            "provider": "google",
            "id_token": googleIdToken
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            let (data, response) = try await URLSession.shared.data(for: request)

            if let httpResponse = response as? HTTPURLResponse {
                print("Supabase token exchange status: \(httpResponse.statusCode)")
            }

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let accessToken = json["access_token"] as? String {
                self.supabaseAccessToken = accessToken
                print("Supabase session obtained")
            } else {
                let responseStr = String(data: data, encoding: .utf8) ?? "no response"
                print("Supabase token exchange failed: \(responseStr)")
            }
        } catch {
            print("Supabase token exchange error: \(error)")
        }
    }


    func signOut() {
        // Disconnect completely to clear cached credentials
        GIDSignIn.sharedInstance.disconnect { _ in
            print("Google Sign-In disconnected")
        }
        GIDSignIn.sharedInstance.signOut()
        currentUser = nil
        idToken = nil
        supabaseAccessToken = nil
        userUuid = nil
    }

    var isSignedIn: Bool {
        currentUser != nil && idToken != nil
    }

    /// Whether realtime is available (requires Supabase token)
    var isRealtimeAvailable: Bool {
        supabaseAccessToken != nil
    }

    var userId: String? {
        currentUser?.id
    }

    var displayName: String {
        currentUser?.displayName ?? "User"
    }

    /// Convert Google ID to deterministic UUID (matches backend format exactly)
    private func googleIdToUuid(_ googleId: String) -> String {
        let data = "google:\(googleId)".data(using: .utf8)!
        let hash = SHA256.hash(data: data)
        let hashBytes = Array(hash)

        // Take first 16 bytes and format as hex string (32 chars)
        let hex = hashBytes.prefix(16).map { String(format: "%02x", $0) }.joined()

        // Format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
        // Backend JS: hex.slice(0,8)-hex.slice(8,12)-4+hex.slice(13,16)-variantByte+hex.slice(18,20)-hex.slice(20,32)
        let segment1 = String(hex.prefix(8))
        let segment2 = String(hex.dropFirst(8).prefix(4))
        let segment3partial = String(hex.dropFirst(13).prefix(3))

        // Parse hex chars 16-17 as byte, apply variant bits
        let hexChars16_17 = String(hex.dropFirst(16).prefix(2))
        let byte8 = UInt8(hexChars16_17, radix: 16) ?? 0
        let variantByte = (byte8 & 0x3f) | 0x80

        let segment4partial = String(hex.dropFirst(18).prefix(2))
        let segment5 = String(hex.dropFirst(20).prefix(12))

        return "\(segment1)-\(segment2)-4\(segment3partial)-\(String(format: "%02x", variantByte))\(segment4partial)-\(segment5)"
    }
}
